pkg/descheduler/strategies/topologyspreadconstraint.go

```diff
@@ -0,0 +1,336 @@
+// whichever number is less.
+//
+// (Note, we will only move as many pods from a domain as possible without bringing it below the ideal average,
+//  and we will not bring any smaller domain above the average)
+// If the diff is within the skew, we move to the next highest domain.
+// If the higher domain can't give any more without falling below the average, we move to the next lowest "high" domain
+//
+// Following this, the above topology domains end up "sorted" as:
+// [5, 5, 5, 5, 5, 5]
+// (assuming even distribution by the scheduler of the evicted pods)
+func balanceDomains(podsForEviction map[*v1.Pod]struct{}, constraint v1.TopologySpreadConstraint, constraintTopologies map[topologyPair][]*v1.Pod, sumPods float64, isEvictable func(*v1.Pod) bool) {
+	idealAvg := sumPods / float64(len(constraintTopologies))
+	sortedDomains := sortDomains(constraintTopologies, isEvictable)
+	// i is the index for belowOrEqualAvg
+	// j is the index for aboveAvg
+	i := 0
+	j := len(sortedDomains) - 1
+	for i < j {
+		// if j has no more to give without falling below the ideal average, move to next aboveAvg
+		if float64(len(sortedDomains[j].pods)) < idealAvg {
+			j--
+		}
+
+		// skew = diff - maxSkew
+		skew := float64(len(sortedDomains[j].pods) - len(sortedDomains[i].pods))
+
+		// if i and j are within the maxSkew of each other, move to next belowOrEqualAvg
+		if int32(skew) <= constraint.MaxSkew {
+			i++
+			continue
+		}
+
+		// the most that can be given from aboveAvg is:
+		// 1. up to half the distance between them
+		// 2. how many it has remaining without falling below the average, or
+		// 3. how many can be added without bringing the smaller domain above the average,
+		// whichever is less
+		// (This is the basic principle of keeping all sizes within ~skew of the average)
+		aboveAvg := math.Ceil(float64(len(sortedDomains[j].pods)) - idealAvg)
+		belowAvg := math.Ceil(idealAvg - float64(len(sortedDomains[i].pods)))
+		smallestDiff := math.Min(aboveAvg, belowAvg)
+		halfSkew := math.Floor(skew / 2)
+		movePods := int(math.Min(smallestDiff, halfSkew))
+		if movePods <= 0 {
+			i++
+			continue
+		}
+
+		// remove pods from the higher topology and add them to the list of pods to be evicted
+		// also (just for tracking), add them to the list of pods in the lower topology
+		aboveToEvict := sortedDomains[j].pods[len(sortedDomains[j].pods)-movePods:]
+		for i := range aboveToEvict {
+			// if the pod has a hard nodeAffinity or nodeSelector, don't bother evicting it as it will just end up back on the same node
+			// however we still account for it "being evicted" so the algorithm can complete
+			if aboveToEvict[i].Spec.NodeSelector != nil ||
+				(aboveToEvict[i].Spec.Affinity != nil &&
+					aboveToEvict[i].Spec.Affinity.NodeAffinity != nil &&
+					aboveToEvict[i].Spec.Affinity.NodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution != nil) {
```

Comment:
I think we should check if pod is evictable here, given we added non-evictable pods to the list.

Reply:
> it will just end up back on the same node

Node selector can target a set of nodes. The same holds for node affinity. What about to check if the node selector or affinity can target at most one node and only then continue?

Reply:
> I think we should check if pod is evictable here, given we added non-evictable pods to the list.

Pods from `podsForEviction` are not evicted until all constraints are processed. You might be running the check multiple times compared to one if the pod participates in multiple constraints.

Reply:
> Node selector can target a set of nodes. The same holds for node affinity. What about to check if the node selector or affinity can target at most one node and only then continue?

That's a good point, I have actually been wondering about if it really makes sense to hard-exclude any pods with affinity/selector. My thinking was different though: the affinity conditions could change on the cluster making the pod perfectly evictable (see our own descheduling strategies). We could actually re-use some helpers from those strategies in informing this decision, but that's toeing the line of conflating 2 different strategies. I think it would be fine though wdyt?

Reply:
@ingvagabund I added a helper here to count the # of nodes this pod could fit on (besides the one it's currently on). If that value is 0, it either only fits on this node or doesn't fit on any node, either way it's not PTS's job to evict it in that case.

I did add a TODO here, because I'm realizing since we don't order the pods with affinity based on how many other nodes they can fit on, there could be enough valid evictable pods with affinity that we don't get to (since this list is created with just `aboveToEvict := sortedDomains[j].pods[len(sortedDomains[j].pods)-movePods:]`, which is efficient but optimistic). 

The alternative would be to either add additional sorting for affinity pods based on other feasible nodes, or loop through all the pods in `sortedDomains[j]` and maintain a counter, and keep trying until count==movePods or k=0